### ISSUE TITLE
Infinite Loop when using Threshold Attack and Pixel Attack

### ISSUE THREAD

**Issue by taylover2016** [Trusted-AI/adversarial-robustness-toolbox#947]
**Describe the bug**
The code seems to enter an infinite loop when I using the Threshold Attack and Pixel Attack.
Some of the code is as follows:

classifier = PyTorchClassifier(
            model=model,
            input_shape=(data.shape),
            nb_classes=40,
            loss=nn.CrossEntropyLoss(),
            device_type="cpu"
        )

"""
    Black-Box Classifier
"""

def predict(data):
    original_predictions = classifier.predict(data)
    return original_predictions

BlackBox = BlackBoxClassifierNeuralNetwork(
    predict=predict,
    input_shape=data.shape,
    nb_classes=40,
    clip_values=(0, 1)
)

"""# ThresholdAttack
attacker = ThresholdAttack(
    classifier=BlackBox,
    verbose=True
)
"""
"""
# PixelAttack
attacker = PixelAttack(
    classifier=BlackBox,
    verbose=True
)
"""

x_test_adv = attacker.generate(
    x=data
    )

# Evaluate the performance
perturbed_predictions = classifier.predict(x_test_adv)

accuracy = np.sum(np.argmax(perturbed_predictions, axis=1) == np.argmax(test_set_y, axis=1)) / test_set_y.shape[0]
print("Accuracy on adversarial test examples: {}%".format(accuracy * 100))

When I turn on the Verbose, it just keeps running and it seems every time
the verbose output is the same.

But the output may vary:

Pixel threshold: 0it [00:00, ?it/s](13_w,27)-aCMA-ES (mu_w=7.8,w_1=22%) in dimension 2679 (seed=279410, Mon Mar  8 12:47:22 2021)
~/opt/anaconda3/envs/pytorch/lib/python3.8/site-packages/cma/utilities/utils.py:343: UserWarning: function_values is not a list of scalars,
                        the first element equals [3979.1602] with non-scalar type <class 'numpy.ndarray'>.
                        Using now ``[v[0] for v in function_values]`` instead (further warnings are suppressed) ()
  warnings.warn(msg + ' (' +
Iterat #Fevals   function value  axis ratio  sigma  min&max std  t[m:s]
    1     27 3.904695800781250e+03 1.0e+00 1.59e+01  2e+01  2e+01 0:00.4
~/opt/anaconda3/envs/pytorch/lib/python3.8/site-packages/cma/utilities/utils.py:343: UserWarning: function_values is not a list of scalars,
                        the first element equals [3972.8545] with non-scalar type <class 'numpy.ndarray'>.
                        Using now ``[v[0] for v in function_values]`` instead (further warnings are suppressed) ()
  warnings.warn(msg + ' (' +
    2     54 3.864472900390625e+03 1.0e+00 1.59e+01  2e+01  2e+01 0:00.7
    3     81 3.807796875000000e+03 1.0e+00 1.58e+01  2e+01  2e+01 0:01.1
   13    351 3.222232666015625e+03 1.0e+00 1.54e+01  2e+01  2e+01 0:04.3
   20    540 2.839827880859375e+03 1.0e+00 1.52e+01  2e+01  2e+01 0:09.0
   32    864 2.208028320312500e+03 1.0e+00 1.48e+01  1e+01  1e+01 0:14.0
Attack Completed :) Earlier than expected
(13_w,27)-aCMA-ES (mu_w=7.8,w_1=22%) in dimension 2679 (seed=271561, Mon Mar  8 12:47:38 2021)
~/opt/anaconda3/envs/pytorch/lib/python3.8/site-packages/cma/utilities/utils.py:343: UserWarning: function_values is not a list of scalars,
                        the first element equals [4029.7139] with non-scalar type <class 'numpy.ndarray'>.
                        Using now ``[v[0] for v in function_values]`` instead (further warnings are suppressed) ()
  warnings.warn(msg + ' (' +
Iterat #Fevals   function value  axis ratio  sigma  min&max std  t[m:s]
    1     27 4.020949462890625e+03 1.0e+00 7.97e+00  8e+00  8e+00 0:00.3
~/opt/anaconda3/envs/pytorch/lib/python3.8/site-packages/cma/utilities/utils.py:343: UserWarning: function_values is not a list of scalars,
                        the first element equals [4061.6743] with non-scalar type <class 'numpy.ndarray'>.
                        Using now ``[v[0] for v in function_values]`` instead (further warnings are suppressed) ()
  warnings.warn(msg + ' (' +
    2     54 4.004011962890625e+03 1.0e+00 7.95e+00  8e+00  8e+00 0:00.6
    3     81 3.957907226562500e+03 1.0e+00 7.92e+00  8e+00  8e+00 0:00.9
   14    378 3.685729980468750e+03 1.0e+00 7.69e+00  8e+00  8e+00 0:04.1
   20    540 3.562143554687500e+03 1.0e+00 7.58e+00  8e+00  8e+00 0:10.0
   37    999 3.240406250000000e+03 1.0e+00 7.33e+00  7e+00  7e+00 0:15.0
   45   1215 3.080075195312500e+03 1.0e+00 7.24e+00  7e+00  7e+00 0:21.1
   58   1566 2.914004150390625e+03 1.0e+00 7.10e+00  7e+00  7e+00 0:28.3
   77   2079 2.680793701171875e+03 1.0e+00 6.95e+00  7e+00  7e+00 0:36.7
   97   2619 2.481010742187500e+03 1.0e+00 6.82e+00  7e+00  7e+00 0:46.0
  100   2700 2.456019042968750e+03 1.0e+00 6.80e+00  7e+00  7e+00 0:46.9
(13_w,27)-aCMA-ES (mu_w=7.8,w_1=22%) in dimension 2679 (seed=236118, Mon Mar  8 12:48:27 2021)
~/opt/anaconda3/envs/pytorch/lib/python3.8/site-packages/cma/utilities/utils.py:343: UserWarning: function_values is not a list of scalars,
                        the first element equals [4045.799] with non-scalar type <class 'numpy.ndarray'>.
                        Using now ``[v[0] for v in function_values]`` instead (further warnings are suppressed) ()
  warnings.warn(msg + ' (' +
Iterat #Fevals   function value  axis ratio  sigma  min&max std  t[m:s]
    1     27 3.975784179687500e+03 1.0e+00 1.20e+01  1e+01  1e+01 0:00.5
~/opt/anaconda3/envs/pytorch/lib/python3.8/site-packages/cma/utilities/utils.py:343: UserWarning: function_values is not a list of scalars,
                        the first element equals [3982.12] with non-scalar type <class 'numpy.ndarray'>.
                        Using now ``[v[0] for v in function_values]`` instead (further warnings are suppressed) ()
  warnings.warn(msg + ' (' +
    2     54 3.924142578125000e+03 1.0e+00 1.19e+01  1e+01  1e+01 0:00.8
    3     81 3.838447998046875e+03 1.0e+00 1.19e+01  1e+01  1e+01 0:01.3
   12    324 3.488703613281250e+03 1.0e+00 1.16e+01  1e+01  1e+01 0:04.5
   20    540 3.184641601562500e+03 1.0e+00 1.14e+01  1e+01  1e+01 0:09.8
   37    999 2.647020019531250e+03 1.0e+00 1.10e+01  1e+01  1e+01 0:15.1
   46   1242 2.415977539062500e+03 1.0e+00 1.09e+01  1e+01  1e+01 0:21.3
Attack Completed :) Earlier than expected
(13_w,27)-aCMA-ES (mu_w=7.8,w_1=22%) in dimension 2679 (seed=250373, Mon Mar  8 12:48:52 2021)
~/opt/anaconda3/envs/pytorch/lib/python3.8/site-packages/cma/utilities/utils.py:343: UserWarning: function_values is not a list of scalars,
                        the first element equals [4068.834] with non-scalar type <class 'numpy.ndarray'>.
                        Using now ``[v[0] for v in function_values]`` instead (further warnings are suppressed) ()
  warnings.warn(msg + ' (' +
Iterat #Fevals   function value  axis ratio  sigma  min&max std  t[m:s]
    1     27 4.014551025390625e+03 1.0e+00 9.97e+00  1e+01  1e+01 0:00.3
/Users/tonyzhang/opt/anaconda3/envs/pytorch/lib/python3.8/site-packages/cma/utilities/utils.py:343: UserWarning: function_values is not a list of scalars,
                        the first element equals [4007.8179] with non-scalar type <class 'numpy.ndarray'>.
                        Using now ``[v[0] for v in function_values]`` instead (further warnings are suppressed) ()
  warnings.warn(msg + ' (' +
    2     54 3.966331542968750e+03 1.0e+00 9.93e+00  1e+01  1e+01 0:00.6
    3     81 3.947932373046875e+03 1.0e+00 9.90e+00  1e+01  1e+01 0:00.9
   14    378 3.607798828125000e+03 1.0e+00 9.61e+00  1e+01  1e+01 0:04.1
   20    540 3.360371093750000e+03 1.0e+00 9.48e+00  9e+00  9e+00 0:09.1
~/opt/anaconda3/envs/pytorch/lib/python3.8/multiprocessing/resource_tracker.py:216: UserWarning: resource_tracker: There appear to be 1 leaked semaphore objects to clean up at shutdown
  warnings.warn('resource_tracker: There appear to be %d '
Process finished with exit code 0

In the end I terminate the program, but last night it could cost up to almost an hour.
I wonder could it be the lack of my computing sources or the warning?

**System information (please complete the following information):**
 - macOS 11.2.2
 - Python 3.8.8
 - ART 1.5.2
 - PyTorch 1.8.0

**Comment by beat-buesser** [comment]
Hi @taylover2016 Thank you very much for exploring ART. Could you please provide a reproducible example in a Google Colab notebook?

**Comment by beat-buesser** [comment]
The `while True` statement has the potential for an infinite loop. I would be interested to take a closer look. Would you be able to share a notebook with a code example with model and data to reproduce this case?

**Comment by taylover2016** [comment]
@beat-buesser Hi, I've provided the necessary code, the dataset and the model together with the params in the latest reply to the issue #943.
Again, Thx for the reply!

**Comment by shashankkotyan** [comment]
Hi, @taylover2016! I happen to see this bug. Can you pass th=10 in the initialiser and see if it still takes a long time. 

I think it is taking notoriously long time to run since it cannot attack your given samples. 
With the default setting of th=None, it tries to find minimal perturbation possible which is a time-expensive task.

**Comment by taylover2016** [comment]
@shashankkotyan Hi, thx for the advice. I'll go and check it out tomorrow.😁

**Comment by beat-buesser** [comment]
Hi @shashankkotyan Thank you for your help!
